### PR TITLE
[rllib] Add option to proceed even if some workers crashed

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds a `ignore_worker_failures` option, which can be set to keep going even if a worker crashes. Since it's pretty hard to track down the source of a RayTaskError in a nice, algorithm-agnostic way, we do this via a more brute force approach:
- If RayTaskError is encountered in a call to train, we health check each worker by calling sample() on it.
- Workers that respond with error on sample() will be blacklisted
- If the number of available workers drops to zero then we raise a permanent error

The number of healthy workers can be tracked via the `num_healthy_workers` metric now reported.

## Related issue number

Closes https://github.com/ray-project/ray/issues/4358

### CODE REVIEW

File: python/ray/rllib/utils/actors.py [hunk 1]
```diff
@@ -53,6 +53,18 @@ def completed_prefetch(self):
                 remaining.append((worker, obj_id))
         self._fetching = remaining
 
+    def reset_evaluators(self, evaluators):
```
Comment (ericl) [inline python/ray/rllib/utils/actors.py]: This is needed since otherwise a blacklisted evaluator might still have old tasks in the TaskPool object.

File: python/ray/rllib/evaluation/sampler.py [hunk 2]
```diff
@@ -171,6 +171,8 @@ def _run(self):
                 queue_putter(item)
 
     def get_data(self):
+        if not self.is_alive():
+            raise RuntimeError("Sampling thread has died")
```
Comment (ericl) [inline python/ray/rllib/evaluation/sampler.py]: This used to hang otherwise.

File: python/ray/rllib/agents/impala/impala.py [hunk 3]
```diff
@@ -116,7 +116,8 @@ def _train(self):
         prev_steps = self.optimizer.num_steps_sampled
         start = time.time()
         self.optimizer.step()
-        while time.time() - start < self.config["min_iter_time_s"]:
+        while (time.time() - start < self.config["min_iter_time_s"]
+               or self.optimizer.num_steps_sampled == prev_steps):
```
Comment (ericl) [inline python/ray/rllib/agents/impala/impala.py]: Sometimes the iteration finished with 0 timesteps otherwise.

**Comment by ericl** [comment]
@flying-mojo interested in trying this out?

File: python/ray/rllib/tests/test_ignore_worker_failure.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import gym
+import unittest
+
+import ray
+from ray.rllib import _register_all
+from ray.rllib.agents.registry import get_agent_class
+from ray.tune.registry import register_env
+
+
+class FaultInjectEnv(gym.Env):
```
Comment (richardliaw) [inline python/ray/rllib/tests/test_ignore_worker_failure.py]: Can you wrap this in a builder? Pytest will hang because with high probability, it won't find this class.

**Review by richardliaw — APPROVED** [review]
Looks good after pytest issue is addressed

File: python/ray/rllib/tests/test_ignore_worker_failure.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/tests/test_ignore_worker_failure.py]: We don't use pytest for rllib
